File: packages/asfdk/src/foundation.ts
import {
  Channel,
  FoundationConfig,
  FoundationMode,
  FoundationResponse,
  HealthCheckResult,
  InteractionType,
  normalizeChannel,
  UserInteraction,
} from './types.js';
import * as toiOtoi from './integration/toi-otoi.js';
import * as sleepwalker from './integration/sleepwalker.js';
import * as rrt from './integration/rrt.js';
import { validateOutput, logSecurityEvent } from './prompt-defense.js';

function componentsForMode(mode: FoundationMode, overrides?: FoundationConfig['components']): {
  toi: boolean;
  swp: boolean;
  rrt: boolean;
} {
  const defaults = {
    [FoundationMode.UNIFIED]: { toi: true, swp: true, rrt: true },
    [FoundationMode.CRISIS_ONLY]: { toi: false, swp: false, rrt: true },
    [FoundationMode.CONTINUITY_ONLY]: { toi: false, swp: true, rrt: false },
    [FoundationMode.FRAMEWORK_ONLY]: { toi: true, swp: false, rrt: false },
    [FoundationMode.DEVELOPMENT]: { toi: true, swp: true, rrt: false },
  };
  const base = defaults[mode];
  // Fail-loud (T16): an unrecognized mode must never silently disable every
  // component — that would look like "governance off" instead of "bad config".
  if (!base) {
    throw new Error(`Unrecognized foundation mode: '${String(mode)}'`);
  }
  return {
    toi: overrides?.toi_otoi_framework ?? base.toi,
    swp: overrides?.sleepwalker_protocol ?? base.swp,
    rrt: overrides?.rrt_advocate ?? base.rrt,
  };
}

/**
 * Central orchestrator for the Solidarity Framework.
 *
 * Routes user interactions to the active Solidarity Framework components
 * (TOI-OTOI, Sleepwalker Protocol, RRT Advocate) according to the
 * configured {@link FoundationMode}.
 *
 * Obtain an instance via {@link createFoundation} rather than constructing directly.
 */
export class NeuroLiftFoundation {
  private readonly config: FoundationConfig;
  private readonly active: ReturnType<typeof componentsForMode>;
  private initialized = false;

  constructor(config: FoundationConfig) {
    this.config = config;
    this.active = componentsForMode(config.mode, config.components);
  }

  /** Marks the foundation as initialized. Called automatically by {@link createFoundation}. */
  async initialize(): Promise<void> {
    this.initialized = true;
  }

  /** Alias for {@link initialize}; ensures the foundation is ready before use. */
  async start(): Promise<void> {
    if (!this.initialized) await this.initialize();
  }

  /**
   * Validates a component result (e.g. RRT) against the output contract.
   * On failure, logs a security event and returns `undefined` so the caller can
   * leave the corresponding `content.*` field unset — invalid output must never
   * reach a {@link FoundationResponse}.
   */
  private validateComponentOutput(result: unknown): unknown {
    if (result === null || result === undefined) return result;
    const outputValid = validateOutput(JSON.stringify(result), { type: 'json' });
    if (outputValid.valid) return result;
    logSecurityEvent({
      type: 'VALIDATION_FAILURE',
      userId: this.config.userId,
      details: `Component output validation failed: ${outputValid.reason}`,
      timestamp: Date.now(),
    });
    return undefined;
  }

  /**
   * Routes a {@link UserInteraction} to the appropriate active components and
   * returns a {@link FoundationResponse} with aggregated content.
   *
   * Security: Input is sanitized and output is validated to prevent prompt injection.
   *
   * - `EMOTIONAL_ASSESSMENT` → Sleepwalker Protocol (+ RRT handoff if crisis indicated)
   * - `PREFERENCE_UPDATE` → TOI-OTOI schema validation
   * - `CRISIS_ALERT` / `EMERGENCY_ESCALATION` → RRT Advocate crisis detection
   * - All other types → empty `componentsInvolved` array with `success: true`
   */
  async processInteraction(interaction: UserInteraction): Promise<FoundationResponse> {
    const components: string[] = [];
    const content: Record<string, unknown> = {};

    // Provenance (D2/D4/D6): resolve the channel from the TOP-LEVEL field only —
    // values nested inside data/context are ignored for trust (anti-spoofing).
    // Absent → 'unknown'; trusted := channel === 'user_input'.
    const channel = normalizeChannel(interaction.channel ?? undefined);
    const trusted = channel === Channel.USER_INPUT;

    // D5 gate-up predicate: untrusted channel AND high-severity crisis signal.
    let highSeverity = false;

    if (this.active.swp && interaction.interactionType === InteractionType.EMOTIONAL_ASSESSMENT) {
      try {
        const input = String(interaction.data?.['text'] ?? '');
        const state = sleepwalker.detectEmotionalState(input, [], channel, this.config.userId);
        // Emotional-path high-severity: explicit crisis flags on the state.
        highSeverity =
          state.explicitSuicidalIdeation || state.selfHarmIndicators || state.inabilityToEnsureSafety;
        content.emotionalState = state;

        if (this.active.rrt && sleepwalker.requiresRrtaHandoff(state)) {
          // Own error boundary so an RRT failure is attributed to rrt_advocate
          // (not sleepwalker) and does not discard the emotional-state result.
          try {
            const rrtResult = this.validateComponentOutput(await rrt.assess(this.config.userId, input, channel));
            if (rrtResult !== undefined) content.rrt = rrtResult;
          } catch (err) {
            content.error = { component: 'rrt_advocate', message: String(err) };
          }
          // Listed whenever attempted (success or failure), consistent with the
          // crisis/emergency route and with sleepwalker_protocol.
          components.push('rrt_advocate');
        }
      } catch (err) {
        content.error = { component: 'sleepwalker_protocol', message: String(err) };
      }
      components.push('sleepwalker_protocol');
    }

    if (this.active.toi && interaction.interactionType === InteractionType.PREFERENCE_UPDATE) {
      content.toiValidation = toiOtoi.validateTOI(interaction.data?.['toi']);
      components.push('toi_otoi_framework');
    }

    if (
      this.active.rrt &&
      (interaction.interactionType === InteractionType.CRISIS_ALERT ||
        interaction.interactionType === InteractionType.EMERGENCY_ESCALATION)
    ) {
      const input = String(interaction.data?.['text'] ?? '');
      // Error boundary: an RRT failure must not abort a crisis/emergency route.
      try {
        const rrtResult = this.validateComponentOutput(await rrt.assess(this.config.userId, input, channel));
        if (rrtResult !== undefined) content.rrt = rrtResult;
        // CRISIS_ALERT is high-severity only on an actual RED/BLACK reading; a
        // failed detection is not evidence of severity, so it does not gate up.
        if (interaction.interactionType === InteractionType.CRISIS_ALERT) {
          const level = (content.rrt as { crisisLevel?: rrt.CrisisLevel } | undefined)?.crisisLevel;
          highSeverity = level === rrt.CrisisLevel.RED || level === rrt.CrisisLevel.BLACK;
        }
      } catch (err) {
        content.error = { component: 'rrt_advocate', message: String(err) };
      }
      components.push('rrt_advocate');
    }

    // D5 high-severity fallbacks:
    // - EMERGENCY_ESCALATION is always high-severity.
    // - CRISIS_ALERT with RRT inactive is high-severity by interaction type
    //   alone (fail-safe: never silently ignored when detection is off).
    if (interaction.interactionType === InteractionType.EMERGENCY_ESCALATION) {
      highSeverity = true;
    }
    if (
      interaction.interactionType === InteractionType.CRISIS_ALERT &&
      !this.active.rrt
    ) {
      highSeverity = true;
    }

    content.channel = channel;
    content.trusted = trusted;
    content.gateUp = !trusted && highSeverity;

    return {
      timestamp: new Date(),
      responseType: interaction.interactionType,
      content,
      componentsInvolved: components,
      success: true,
    };
  }

  /**
   * Assesses the emotional state of a free-text input via the Sleepwalker Protocol.
   * Returns `null` when Sleepwalker is not active for the current mode.
   *
   * Channel provenance (D4) is recorded additively on the returned assessment —
   * no envelope — so existing consumers see the same shape plus `channel`,
   * `trusted`, and `gateUp` properties. Absent channel → `unknown`/untrusted.
   *
   * @param input - Free-text user input to assess.
   * @param _context - Reserved for future context enrichment; currently unused.
   * @param channel - Optional channel the interaction arrived on; absent → `unknown`.
   */
  async assessEmotionalState(
    input: string,
    _context?: Record<string, unknown>,
    channel?: Channel,
  ): Promise<unknown> {
    if (!this.active.swp) return null;
    const resolved = normalizeChannel(channel ?? undefined);
    const trusted = resolved === Channel.USER_INPUT;
    const result = sleepwalker.assessInteraction(input) as Record<string, unknown>;
    // assessInteraction nests the state; read the crisis flags from it (falling
    // back to the top level for flat-shaped callers).
    const inner = (result.emotionalState ?? result) as Record<string, unknown>;
    const highSeverity = Boolean(
      inner.explicitSuicidalIdeation ||
        inner.selfHarmIndicators ||
        inner.inabilityToEnsureSafety,
    );
    result.channel = resolved;
    result.trusted = trusted;
    result.gateUp = !trusted && highSeverity;
    return result;
  }

  /**
   * Validates a preference object against the TOI schema and throws if invalid.
   * No-op when TOI-OTOI is not active for the current mode.
   *
   * @throws {Error} If the preference object fails TOI schema validation.
   */
  async updatePreferences(prefs: Record<string, unknown>): Promise<void> {
    if (this.active.toi) {
      const result = toiOtoi.validateTOI(prefs);
      if (!result.valid) {
        throw new Error('TOI validation failed: ' + JSON.stringify(result.errors));
      }
    }
  }

  /** Returns the current mode, userId, initialization state, and per-component status. */
  getSystemStatus(): Record<string, unknown> {
    return {
      mode: this.config.mode,
      userId: this.config.userId,
      initialized: this.initialized,
      components: {
        toi_otoi_framework: this.active.toi ? toiOtoi.getStatus() : { active: false, mode: 'disabled' },
        sleepwalker_protocol: this.active.swp ? sleepwalker.getStatus() : { active: false, mode: 'disabled' },
        rrt_advocate: this.active.rrt ? rrt.getStatus() : { active: false, mode: 'disabled' },
      },
    };
  }

  /**
   * Returns a structured health report for all components, reflecting which are
   * active for the current {@link FoundationMode}.
   */
  async healthCheck(): Promise<HealthCheckResult> {
    return {
      healthy: true,
      timestamp: new Date(),
      components: {
        toi_otoi_framework: this.active.toi
          ? { active: true, mode: 'toi-otoi-validation' }
          : { active: false, mode: 'disabled' },
        sleepwalker_protocol: this.active.swp
          ? { active: true, mode: 'emotional-continuity' }
          : { active: false, mode: 'disabled' },
        rrt_advocate: this.active.rrt
          ? { active: true, mode: 'crisis-detection' }
          : { active: false, mode: 'disabled' },
      },
    };
  }

  /** Resets Sleepwalker and RRT Advocate state and marks the foundation as uninitialized. */
  async shutdown(): Promise<void> {
    sleepwalker.reset();
    rrt.reset(this.config.userId);
    this.initialized = false;
  }
}

File: src/asfdk/foundation.py
"""The :class:`NeuroLiftFoundation` orchestrator, ported from
``@neurolift-technologies/asfdk`` (``src/foundation.ts``).
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Union

try:  # ``nlt-toi`` >= 1.0.0 ships this alias; keep annotations meaningful either way.
    from nlt_toi import ToiDocument
except ImportError:  # pragma: no cover - defensive
    ToiDocument = Dict[str, Any]  # type: ignore[misc,assignment]

from .integration import rrt, sleepwalker, toi_otoi
from .toi_bootstrap import generator_source, get_generator_class
from .types import (
    Channel,
    ComponentStatus,
    FoundationComponents,
    FoundationConfig,
    FoundationMode,
    FoundationResponse,
    HealthCheckResult,
    InteractionType,
    UserInteraction,
    normalize_channel,
)


@dataclass
class _ActiveComponents:
    toi: bool
    swp: bool
    rrt: bool


def _components_for_mode(
    mode: FoundationMode,
    overrides: Optional[FoundationComponents] = None,
) -> _ActiveComponents:
    defaults: Dict[FoundationMode, _ActiveComponents] = {
        FoundationMode.UNIFIED: _ActiveComponents(toi=True, swp=True, rrt=True),
        FoundationMode.CRISIS_ONLY: _ActiveComponents(toi=False, swp=False, rrt=True),
        FoundationMode.CONTINUITY_ONLY: _ActiveComponents(toi=False, swp=True, rrt=False),
        FoundationMode.FRAMEWORK_ONLY: _ActiveComponents(toi=True, swp=False, rrt=False),
        FoundationMode.DEVELOPMENT: _ActiveComponents(toi=True, swp=True, rrt=False),
    }
    base = defaults.get(mode, _ActiveComponents(toi=False, swp=False, rrt=False))

    def pick(override: Optional[bool], fallback: bool) -> bool:
        # Mirror TS ``override ?? fallback`` — None means "not provided".
        return fallback if override is None else override

    return _ActiveComponents(
        toi=pick(overrides.toi_otoi_framework if overrides else None, base.toi),
        swp=pick(overrides.sleepwalker_protocol if overrides else None, base.swp),
        rrt=pick(overrides.rrt_advocate if overrides else None, base.rrt),
    )


class NeuroLiftFoundation:
    """Central orchestrator for the Solidarity Framework.

    Routes user interactions to the active Solidarity Framework components
    (TOI-OTOI, Sleepwalker Protocol, RRT Advocate) according to the configured
    :class:`~asfdk.types.FoundationMode`.

    Obtain an instance via :func:`asfdk.create_foundation` rather than
    constructing directly.
    """

    def __init__(self, config: FoundationConfig) -> None:
        self._config = config
        self._active = _components_for_mode(config.mode, config.components)
        self._initialized = False
        #: The foundation's active ``.toi`` document, generated before any
        #: component activates (see :meth:`initialize`). ``None`` until
        #: initialization succeeds.
        self._toi_document: Optional[ToiDocument] = None

    async def initialize(self) -> None:
        """Bootstrap the foundation.

        The TOI generator runs *first*: the foundation's active document is
        generated/validated from the configured source (or privacy-first
        defaults) and stored before any component becomes active. A failure to
        generate a conforming document raises and leaves the foundation
        uninitialized (fail-loud), so components are never activated against a
        broken or absent TOI.
        """
        self._toi_document = self._generate_toi()
        self._initialized = True

    def _generate_toi(self) -> ToiDocument:
        """Run the ``toi-generator`` logic over the configured source.

        The authoring helper is the pillar's own ``TOIDocumentGenerator`` when the
        installed ``nlt-toi`` provides it, otherwise the local mirror in
        :mod:`asfdk.toi_bootstrap` (which delegates validation to the pillar's
        canonical schema). :func:`asfdk.toi_bootstrap.generator_source` reports
        which one is in use.

        Source resolution:
        - ``None``  → privacy-first document generated from defaults
        - ``str``   → path to a ``.toi``/``.json`` file, parsed then regenerated
        - ``dict``  → partial preferences or a full document, merged over defaults

        Every path validates through the canonical schema before returning, so
        an invalid source raises here rather than activating a broken TOI.
        """
        generator = get_generator_class()
        source = self._config.toi
        if source is None:
            return generator.from_defaults("anonymous").document
        if isinstance(source, dict):
            return generator.from_dict(source).document
        if isinstance(source, str):
            with open(source, encoding="utf-8") as handle:
                raw = json.load(handle)
            if not isinstance(raw, dict):
                raise TypeError("a .toi source file must contain a JSON object")
            return generator.from_dict(raw).document
        raise TypeError(
            "toi must be a preferences dict, a path to a .toi/.json file, or None; "
            f"got {type(source).__name__}"
        )

    def get_active_toi(self) -> Optional[ToiDocument]:
        """Return the foundation's active TOI document, or ``None`` before
        :meth:`initialize` succeeds."""
        return self._toi_document

    async def start(self) -> None:
        """Alias for :meth:`initialize`; ensures the foundation is ready before use."""
        if not self._initialized:
            await self.initialize()

    async def process_interaction(
        self, interaction: UserInteraction
    ) -> FoundationResponse:
        """Route a :class:`~asfdk.types.UserInteraction` to the appropriate active
        components and return a :class:`~asfdk.types.FoundationResponse` with
        aggregated content.

        - ``EMOTIONAL_ASSESSMENT`` → Sleepwalker Protocol (+ RRT handoff if crisis indicated)
        - ``PREFERENCE_UPDATE`` → TOI-OTOI schema validation
        - ``CRISIS_ALERT`` / ``EMERGENCY_ESCALATION`` → RRT Advocate crisis detection
        - All other types → empty ``components_involved`` list with ``success=True``
        """
        components: List[str] = []
        content: Dict[str, Any] = {}

        # Provenance (D2/D4/D6): resolve the channel from the TOP-LEVEL field
        # only — values nested inside data/context are ignored for trust
        # (anti-spoofing). Absent → ``unknown``; trusted := channel == user_input.
        channel = normalize_channel(interaction.channel)
        trusted = channel == Channel.USER_INPUT
        # D5 gate-up predicate: untrusted channel AND high-severity crisis signal.
        high_severity = False
        if (
            self._active.swp
            and interaction.interaction_type == InteractionType.EMOTIONAL_ASSESSMENT
        ):
            try:
                text_val = (interaction.data or {}).get("text")
                user_input = str(text_val) if text_val is not None else ""
                state = sleepwalker.detect_emotional_state(
                    user_input, [], channel, self._config.user_id
                )
                # Emotional-path high-severity: explicit crisis flags on the state.
                high_severity = bool(
                    state.explicit_suicidal_ideation
                    or state.self_harm_indicators
                    or state.inability_to_ensure_safety
                )
                content["emotionalState"] = state

                if self._active.rrt and sleepwalker.requires_rrta_handoff(state):
                    # Own error boundary so an RRT failure is attributed to
                    # rrt_advocate (not sleepwalker) and does not discard the
                    # emotional-state result.
                    try:
                        content["rrt"] = await rrt.assess(
                            self._config.user_id, user_input, channel
                        )
                    except Exception as err:  # noqa: BLE001
                        content["error"] = {
                            "component": "rrt_advocate",
                            "message": str(err),
                        }
                    # Listed whenever attempted (success or failure), consistent
                    # with the crisis/emergency route and with sleepwalker_protocol.
                    components.append("rrt_advocate")
            except Exception as err:  # noqa: BLE001
                content["error"] = {
                    "component": "sleepwalker_protocol",
                    "message": str(err),
                }
            components.append("sleepwalker_protocol")

        if (
            self._active.toi
            and interaction.interaction_type == InteractionType.PREFERENCE_UPDATE
        ):
            content["toiValidation"] = toi_otoi.validate_toi(
                (interaction.data or {}).get("toi")
            )
            components.append("toi_otoi_framework")

        if self._active.rrt and interaction.interaction_type in (
            InteractionType.CRISIS_ALERT,
            InteractionType.EMERGENCY_ESCALATION,
        ):
            text_val = (interaction.data or {}).get("text")
            user_input = str(text_val) if text_val is not None else ""
            # Error boundary: an RRT failure must not abort a crisis/emergency route.
            try:
                content["rrt"] = await rrt.assess(
                    self._config.user_id, user_input, channel
                )
                # CRISIS_ALERT is high-severity only on an actual RED/BLACK
                # reading; a failed detection is not evidence of severity, so it
                # does not gate up.
                if interaction.interaction_type == InteractionType.CRISIS_ALERT:
                    level = getattr(content.get("rrt"), "crisis_level", None)
                    high_severity = level in (
                        rrt.CrisisLevel.RED,
                        rrt.CrisisLevel.BLACK,
                    )
            except Exception as err:  # noqa: BLE001
                content["error"] = {"component": "rrt_advocate", "message": str(err)}
            components.append("rrt_advocate")

        # D5 high-severity fallbacks:
        # - EMERGENCY_ESCALATION is always high-severity.
        # - CRISIS_ALERT with RRT inactive is high-severity by interaction type
        #   alone (fail-safe: never silently ignored when detection is off).
        if interaction.interaction_type == InteractionType.EMERGENCY_ESCALATION:
            high_severity = True
        if (
            interaction.interaction_type == InteractionType.CRISIS_ALERT
            and not self._active.rrt
        ):
            high_severity = True

        # D5 gate-up (Enforce): an untrusted channel carrying a high-severity
        # signal is the one combination that escalates rather than degrades.
        content["channel"] = channel
        content["trusted"] = trusted
        content["gateUp"] = not trusted and high_severity

        return FoundationResponse(
            timestamp=datetime.now(timezone.utc),
            response_type=interaction.interaction_type.value
            if isinstance(interaction.interaction_type, InteractionType)
            else str(interaction.interaction_type),
            content=content,
            components_involved=components,
            success=True,
        )

    async def assess_emotional_state(
        self,
        input: str,
        _context: Optional[Dict[str, Any]] = None,
        channel: Optional[Channel] = None,
    ) -> Any:
        """Assess the emotional state of a free-text input via the Sleepwalker
        Protocol. Returns ``None`` when Sleepwalker is not active for the current
        mode.

        Channel provenance (D4): the resolved channel and its derived ``trusted``
        flag are recorded additively on the returned assessment — no envelope —
        so existing consumers see the same shape plus ``channel``, ``trusted``,
        and ``gateUp`` properties. Absent channel → ``unknown``/untrusted.

        :param input: Free-text user input to assess.
        :param _context: Reserved for future context enrichment; currently unused.
        :param channel: Optional channel the interaction arrived on; absent → ``unknown``.
        """
        if not self._active.swp:
            return None
        resolved = normalize_channel(channel)
        trusted = resolved == Channel.USER_INPUT

        def _get(obj: Any, name: str, default: Any = None) -> Any:
            value = obj.get(name, default) if isinstance(obj, dict) else getattr(obj, name, default)
            return default if value is None else value

        result = sleepwalker.assess_interaction(input)
        # assessInteraction nests the state; read the crisis flags from it
        # (falling back to the top level for flat-shaped callers).
        inner = _get(result, "emotional_state", None) or _get(
            result, "emotionalState", None
        )
        if inner is None:
            inner = result
        high_severity = bool(
            _get(inner, "explicit_suicidal_ideation", False)
            or _get(inner, "explicitSuicidalIdeation", False)
            or _get(inner, "self_harm_indicators", False)
            or _get(inner, "selfHarmIndicators", False)
            or _get(inner, "inability_to_ensure_safety", False)
            or _get(inner, "inabilityToEnsureSafety", False)
        )
        gate_up = not trusted and high_severity
        if isinstance(result, dict):
            result["channel"] = resolved
            result["trusted"] = trusted
            result["gateUp"] = gate_up
        else:
            result.channel = resolved
            result.trusted = trusted
            result.gateUp = gate_up
        return result

    async def update_preferences(self, prefs: Dict[str, Any]) -> None:
        """Validate a preference object against the TOI schema and raise if
        invalid. No-op when TOI-OTOI is not active for the current mode.

        :raises ValueError: If the preference object fails TOI schema validation.
        """
        if self._active.toi:
            result = toi_otoi.validate_toi(prefs)
            if not result.valid:
                errors = [
                    {"message": e.message, "path": e.path, "code": e.code}
                    for e in (result.errors or [])
                ]
                raise ValueError(
                    "TOI validation failed: " + json.dumps(errors, separators=(",", ":"))
                )

    def get_system_status(self) -> Dict[str, Any]:
        """Return the current mode, user_id, initialization state, and
        per-component status.
        """
        return {
            "mode": self._config.mode,
            "userId": self._config.user_id,
            "initialized": self._initialized,
            "toi": {
                "generated": self._toi_document is not None,
                # ``"nlt_toi"`` when the pillar supplies the authoring helper,
                # otherwise ``"asfdk-fallback"`` — the fallback is never silent.
                "generator": generator_source(),
                "document": self._toi_document,
            },
            "components": {
                "toi_otoi_framework": toi_otoi.get_status()
                if self._active.toi
                else {"active": False, "mode": "disabled"},
                "sleepwalker_protocol": sleepwalker.get_status()
                if self._active.swp
                else {"active": False, "mode": "disabled"},
                "rrt_advocate": rrt.get_status()
                if self._active.rrt
                else {"active": False, "mode": "disabled"},
            },
        }

    async def health_check(self) -> HealthCheckResult:
        """Return a structured health report for all components, reflecting which
        are active for the current :class:`~asfdk.types.FoundationMode`.
        """
        return HealthCheckResult(
            healthy=True,
            timestamp=datetime.now(timezone.utc),
            components={
                "toi_otoi_framework": ComponentStatus(active=True, mode="toi-otoi-validation")
                if self._active.toi
                else ComponentStatus(active=False, mode="disabled"),
                "sleepwalker_protocol": ComponentStatus(active=True, mode="emotional-continuity")
                if self._active.swp
                else ComponentStatus(active=False, mode="disabled"),
                "rrt_advocate": ComponentStatus(active=True, mode="crisis-detection")
                if self._active.rrt
                else ComponentStatus(active=False, mode="disabled"),
            },
        )

    async def shutdown(self) -> None:
        """Reset Sleepwalker and RRT Advocate state and mark the foundation as
        uninitialized.
        """
        sleepwalker.reset()
        rrt.reset(self._config.user_id)
        self._initialized = False


__all__ = ["NeuroLiftFoundation"]
